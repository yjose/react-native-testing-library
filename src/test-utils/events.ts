interface EventEntry {
  id: number;
  name: string;
  nativeEvent: any;
}

export function createEventToolkit() {
  let id = 0;
  const events: EventEntry[] = [];
  const handleEvent = (name: string) => {
    return (event: any) => {
      events.push({
        id,
        name,
        nativeEvent: event?.nativeEvent,
      });

      id += 1;
    };
  };

  return { events, handleEvent };
}
