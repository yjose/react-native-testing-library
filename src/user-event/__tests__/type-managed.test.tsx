import * as React from 'react';
import { TextInput } from 'react-native';
import { createEventToolkit } from '../../test-utils/events';
import { render } from '../..';
import { userEvent } from '..';

interface ManagedTextInputProps {
  handleEvent: (name: string) => (event: any) => void;
  fixedValue?: string;
}

function ManagedTextInput({ handleEvent, fixedValue }: ManagedTextInputProps) {
  const [value, setValue] = React.useState('');

  const handleChangeText = (text: string) => {
    handleEvent('changeText')(text);
    setValue(text);
  };

  return (
    <TextInput
      testID="input"
      value={fixedValue ?? value}
      onChangeText={handleChangeText}
      onFocus={handleEvent('focus')}
      onBlur={handleEvent('blur')}
      onPressIn={handleEvent('pressIn')}
      onPressOut={handleEvent('pressOut')}
      onChange={handleEvent('change')}
      onTextInput={handleEvent('textInput')}
      onKeyPress={handleEvent('keyPress')}
      onSelectionChange={handleEvent('selectionChange')}
      onSubmitEditing={handleEvent('onSubmitEditing')}
      onContentSizeChange={handleEvent('contentSizeChange')}
    />
  );
}

test('userEvent.type on TextInput', async () => {
  const { events, handleEvent } = createEventToolkit();
  const screen = render(<ManagedTextInput handleEvent={handleEvent} />);
  await userEvent.type(screen.getByTestId('input'), 'Hello');

  const entryEvents = events.slice(0, 3);
  const exitEvents = events.slice(events.length - 2);
  const typingEvents = events.slice(
    entryEvents.length,
    events.length - exitEvents.length
  );

  expect(entryEvents).toMatchSnapshot('entry events');
  expect(typingEvents).toMatchSnapshot('typing events');
  expect(exitEvents).toMatchSnapshot('exit events');
});

test('userEvent.type on rejecting TextInput', async () => {
  const { events, handleEvent } = createEventToolkit();
  const screen = render(<ManagedTextInput handleEvent={handleEvent} />);
  await userEvent.type(screen.getByTestId('input'), 'Hello');

  const entryEvents = events.slice(0, 3);
  const exitEvents = events.slice(events.length - 2);
  const typingEvents = events.slice(
    entryEvents.length,
    events.length - exitEvents.length
  );

  expect(entryEvents).toMatchSnapshot('entry events');
  expect(typingEvents).toMatchSnapshot('typing events');
  expect(exitEvents).toMatchSnapshot('exit events');
});
