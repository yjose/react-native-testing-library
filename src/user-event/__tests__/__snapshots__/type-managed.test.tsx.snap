// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userEvent.type on TextInput: entry events 1`] = `
[
  {
    "id": 0,
    "name": "pressIn",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759917694,
      "touches": [],
    },
  },
  {
    "id": 1,
    "name": "focus",
    "nativeEvent": {
      "target": 0,
    },
  },
  {
    "id": 2,
    "name": "pressOut",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759917698,
      "touches": [],
    },
  },
]
`;

exports[`userEvent.type on TextInput: exit events 1`] = `
[
  {
    "id": 28,
    "name": "onSubmitEditing",
    "nativeEvent": {
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 29,
    "name": "blur",
    "nativeEvent": {
      "target": 0,
    },
  },
]
`;

exports[`userEvent.type on TextInput: typing events 1`] = `
[
  {
    "id": 3,
    "name": "keyPress",
    "nativeEvent": {
      "key": "H",
    },
  },
  {
    "id": 4,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "",
      "range": {
        "end": 1,
        "start": 1,
      },
      "target": 0,
      "text": "H",
    },
  },
  {
    "id": 5,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "H",
    },
  },
  {
    "id": 6,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 7,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 1,
        "start": 1,
      },
    },
  },
  {
    "id": 8,
    "name": "keyPress",
    "nativeEvent": {
      "key": "e",
    },
  },
  {
    "id": 9,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "H",
      "range": {
        "end": 2,
        "start": 2,
      },
      "target": 0,
      "text": "He",
    },
  },
  {
    "id": 10,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "He",
    },
  },
  {
    "id": 11,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 12,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 2,
        "start": 2,
      },
    },
  },
  {
    "id": 13,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 14,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "He",
      "range": {
        "end": 3,
        "start": 3,
      },
      "target": 0,
      "text": "Hel",
    },
  },
  {
    "id": 15,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hel",
    },
  },
  {
    "id": 16,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 17,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 3,
        "start": 3,
      },
    },
  },
  {
    "id": 18,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 19,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hel",
      "range": {
        "end": 4,
        "start": 4,
      },
      "target": 0,
      "text": "Hell",
    },
  },
  {
    "id": 20,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hell",
    },
  },
  {
    "id": 21,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 22,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 4,
        "start": 4,
      },
    },
  },
  {
    "id": 23,
    "name": "keyPress",
    "nativeEvent": {
      "key": "o",
    },
  },
  {
    "id": 24,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hell",
      "range": {
        "end": 5,
        "start": 5,
      },
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 25,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 26,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 27,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 5,
        "start": 5,
      },
    },
  },
]
`;

exports[`userEvent.type on rejecting TextInput: entry events 1`] = `
[
  {
    "id": 0,
    "name": "pressIn",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759917724,
      "touches": [],
    },
  },
  {
    "id": 1,
    "name": "focus",
    "nativeEvent": {
      "target": 0,
    },
  },
  {
    "id": 2,
    "name": "pressOut",
    "nativeEvent": {
      "changedTouches": [],
      "identifier": 0,
      "locationX": 0,
      "locationY": 0,
      "pageX": 0,
      "pageY": 0,
      "target": 0,
      "timestamp": 1682759917726,
      "touches": [],
    },
  },
]
`;

exports[`userEvent.type on rejecting TextInput: exit events 1`] = `
[
  {
    "id": 28,
    "name": "onSubmitEditing",
    "nativeEvent": {
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 29,
    "name": "blur",
    "nativeEvent": {
      "target": 0,
    },
  },
]
`;

exports[`userEvent.type on rejecting TextInput: typing events 1`] = `
[
  {
    "id": 3,
    "name": "keyPress",
    "nativeEvent": {
      "key": "H",
    },
  },
  {
    "id": 4,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "",
      "range": {
        "end": 1,
        "start": 1,
      },
      "target": 0,
      "text": "H",
    },
  },
  {
    "id": 5,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "H",
    },
  },
  {
    "id": 6,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 7,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 1,
        "start": 1,
      },
    },
  },
  {
    "id": 8,
    "name": "keyPress",
    "nativeEvent": {
      "key": "e",
    },
  },
  {
    "id": 9,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "H",
      "range": {
        "end": 2,
        "start": 2,
      },
      "target": 0,
      "text": "He",
    },
  },
  {
    "id": 10,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "He",
    },
  },
  {
    "id": 11,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 12,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 2,
        "start": 2,
      },
    },
  },
  {
    "id": 13,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 14,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "He",
      "range": {
        "end": 3,
        "start": 3,
      },
      "target": 0,
      "text": "Hel",
    },
  },
  {
    "id": 15,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hel",
    },
  },
  {
    "id": 16,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 17,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 3,
        "start": 3,
      },
    },
  },
  {
    "id": 18,
    "name": "keyPress",
    "nativeEvent": {
      "key": "l",
    },
  },
  {
    "id": 19,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hel",
      "range": {
        "end": 4,
        "start": 4,
      },
      "target": 0,
      "text": "Hell",
    },
  },
  {
    "id": 20,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hell",
    },
  },
  {
    "id": 21,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 22,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 4,
        "start": 4,
      },
    },
  },
  {
    "id": 23,
    "name": "keyPress",
    "nativeEvent": {
      "key": "o",
    },
  },
  {
    "id": 24,
    "name": "textInput",
    "nativeEvent": {
      "previousText": "Hell",
      "range": {
        "end": 5,
        "start": 5,
      },
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 25,
    "name": "change",
    "nativeEvent": {
      "target": 0,
      "text": "Hello",
    },
  },
  {
    "id": 26,
    "name": "changeText",
    "nativeEvent": undefined,
  },
  {
    "id": 27,
    "name": "selectionChange",
    "nativeEvent": {
      "selection": {
        "end": 5,
        "start": 5,
      },
    },
  },
]
`;
